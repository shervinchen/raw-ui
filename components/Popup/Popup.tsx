import React, {
  FC,
  MutableRefObject,
  PropsWithChildren,
  useEffect,
  useState,
  MouseEvent,
} from "react";
import { createPortal } from "react-dom";
import { PopupProps } from "./Popup.types";
import {
  useClickAnyWhere,
  useMutationObserver,
  usePortal,
  useResize,
} from "../utils/hooks";

interface PopupRect {
  top: number;
  left: number;
  width: number;
}

const defaultPopupRect: PopupRect = {
  top: 0,
  left: 0,
  width: 0,
};

const getElementOffset = (element?: HTMLElement | null) => {
  if (!element) return { offsetTop: 0, offsetLeft: 0 };
  const { top, left } = element.getBoundingClientRect();
  return { offsetTop: top, offsetLeft: left };
};

const computePopupRect = (
  targetRef?: MutableRefObject<HTMLElement | null>,
  getContainer?: () => HTMLElement | null
): PopupRect => {
  const targetRect = targetRef?.current?.getBoundingClientRect() ?? null;
  const bodyRect = document.body.getBoundingClientRect();
  const container = getContainer?.() ?? null;
  const { offsetTop, offsetLeft } = getElementOffset(container);

  if (!targetRect) return defaultPopupRect;

  return {
    top: container
      ? targetRect.bottom + container.scrollTop - offsetTop
      : targetRect.bottom - bodyRect.top,
    left: container
      ? targetRect.left + container.scrollLeft - offsetLeft
      : targetRect.left - bodyRect.left,
    width: targetRect.width || targetRect.right - targetRect.left,
  };
};

const Popup: FC<PropsWithChildren<PopupProps>> = ({
  name,
  visible,
  targetRef,
  getPopupContainer,
  children,
}) => {
  const portal = usePortal(name, getPopupContainer);
  const [popupRect, setPopupRect] = useState<PopupRect>(defaultPopupRect);

  const clickHandler = (event: MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    event.nativeEvent.stopImmediatePropagation();
    event.preventDefault();
  };

  const mouseDownHandler = (event: MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
  };

  const updatePopupRect = () => {
    const rect = computePopupRect(targetRef, getPopupContainer);
    setPopupRect(rect);
  };

  useResize(updatePopupRect);

  useClickAnyWhere(() => {
    const { top, left } = computePopupRect(targetRef, getPopupContainer);
    const shouldUpdate = top !== popupRect.top || left !== popupRect.left;
    if (shouldUpdate) updatePopupRect();
  });

  useMutationObserver(targetRef, updatePopupRect);

  useEffect(() => {
    if (!targetRef?.current) return;
    targetRef.current.addEventListener("mouseenter", updatePopupRect);
    return () => {
      if (!targetRef?.current) return;
      targetRef.current.removeEventListener("mouseenter", updatePopupRect);
    };
  }, [targetRef]);

  if (!targetRef) return null;
  if (!portal) return null;

  return createPortal(
    visible ? (
      <div
        className="raw-popup"
        onClick={clickHandler}
        onMouseDown={mouseDownHandler}
      >
        {children}
        <style jsx>{`
          .raw-popup {
            position: absolute;
            top: ${popupRect.top + 2}px;
            left: ${popupRect.left}px;
            width: ${popupRect.width}px;
          }
        `}</style>
      </div>
    ) : null,
    portal
  );
};

export default Popup;
