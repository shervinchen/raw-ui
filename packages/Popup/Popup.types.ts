import { MutableRefObject } from 'react';

export interface PopupProps {
  name: string;
  visible: boolean;
  targetRef?: MutableRefObject<HTMLElement | null>;
  getPopupContainer?: () => HTMLElement | null;
}

export interface PopupRect {
  top: number;
  left: number;
  width: number;
}
