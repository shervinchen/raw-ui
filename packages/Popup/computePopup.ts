import { MutableRefObject } from 'react';
import {
  PopupArrowOffset,
  PopupArrowPosition,
  PopupCoordinates,
  PopupPlacement,
  PopupPosition,
} from './Popup.types';

export const arrowSize = 8;

export const getTargetRect = (
  targetRef?: MutableRefObject<HTMLElement | null>
) => {
  const targetRect = targetRef?.current?.getBoundingClientRect() ?? null;
  const width = targetRect
    ? targetRect.width || targetRect.right - targetRect.left
    : 0;
  const height = targetRect
    ? targetRect.height || targetRect.bottom - targetRect.top
    : 0;

  return {
    width,
    height,
  };
};

const getElementOffset = (element?: HTMLElement | null) => {
  if (!element) return { offsetTop: 0, offsetLeft: 0 };
  const { top, left } = element.getBoundingClientRect();
  return { offsetTop: top, offsetLeft: left };
};

export const computePopupCoordinates = (
  targetRef?: MutableRefObject<HTMLElement | null>,
  getContainer?: () => HTMLElement | null
): PopupCoordinates => {
  const targetRect = targetRef?.current?.getBoundingClientRect() ?? null;
  const bodyRect = document.body.getBoundingClientRect();
  const container = getContainer?.() ?? null;
  const { offsetTop, offsetLeft } = getElementOffset(container);

  if (!targetRect) {
    return {
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    };
  }

  return {
    top: container
      ? targetRect.top + container.scrollTop - offsetTop
      : targetRect.top - bodyRect.top,
    bottom: container
      ? targetRect.bottom + container.scrollTop - offsetTop
      : targetRect.bottom - bodyRect.top,
    left: container
      ? targetRect.left + container.scrollLeft - offsetLeft
      : targetRect.left - bodyRect.left,
    right: container
      ? targetRect.right + container.scrollLeft - offsetLeft
      : targetRect.right - bodyRect.left,
  };
};

export const computePopupPosition = (
  placement: PopupPlacement,
  targetRef?: MutableRefObject<HTMLElement | null>,
  getPopupContainer?: () => HTMLElement | null
) => {
  const { top, bottom, left, right } = computePopupCoordinates(
    targetRef,
    getPopupContainer
  );
  const { width, height } = getTargetRect(targetRef);

  const placements: {
    [key in PopupPlacement]: PopupPosition;
  } = {
    top: {
      top: top - arrowSize,
      left: left + width / 2,
      transform: 'translate(-50%, -100%)',
    },
    topLeft: {
      top: top - arrowSize,
      left,
      transform: 'translate(0, -100%)',
    },
    topRight: {
      top: top - arrowSize,
      left: left + width,
      transform: 'translate(-100%, -100%)',
    },
    bottom: {
      top: bottom + arrowSize,
      left: left + width / 2,
      transform: 'translate(-50%, 0)',
    },
    bottomLeft: {
      top: bottom + arrowSize,
      left,
      transform: 'translate(0, 0)',
    },
    bottomRight: {
      top: bottom + arrowSize,
      left: left + width,
      transform: 'translate(-100%, 0)',
    },
    left: {
      top: top + height / 2,
      left: left - arrowSize,
      transform: 'translate(-100%, -50%)',
    },
    leftTop: {
      top,
      left: left - arrowSize,
      transform: 'translate(-100%, 0)',
    },
    leftBottom: {
      top: top + height,
      left: left - arrowSize,
      transform: 'translate(-100%, -100%)',
    },
    right: {
      top: top + height / 2,
      left: right + arrowSize,
      transform: 'translate(0, -50%)',
    },
    rightTop: {
      top,
      left: right + arrowSize,
      transform: 'translate(0, 0)',
    },
    rightBottom: {
      top: top + height,
      left: right + arrowSize,
      transform: 'translate(0, -100%)',
    },
  };

  return placements[placement] || placements['top'];
};

export const computePopupArrowPosition = (
  arrowOffset: PopupArrowOffset,
  arrowDistance: string,
  placement: PopupPlacement
) => {
  const placements: {
    [key in PopupPlacement]: PopupArrowPosition;
  } = {
    top: {
      top: 'auto',
      right: 'auto',
      left: '50%',
      bottom: arrowDistance,
      transform: 'translate(-50%, 100%) rotate(45deg)',
    },
    topLeft: {
      top: 'auto',
      right: 'auto',
      left: `${arrowOffset.x}`,
      bottom: arrowDistance,
      transform: 'translate(0, 100%) rotate(45deg)',
    },
    topRight: {
      top: 'auto',
      right: `${arrowOffset.x}`,
      left: 'auto',
      bottom: arrowDistance,
      transform: 'translate(0, 100%) rotate(45deg)',
    },
    bottom: {
      top: arrowDistance,
      right: 'auto',
      left: '50%',
      bottom: 'auto',
      transform: 'translate(-50%, -100%) rotate(45deg)',
    },
    bottomLeft: {
      top: arrowDistance,
      right: 'auto',
      left: `${arrowOffset.x}`,
      bottom: 'auto',
      transform: 'translate(0, -100%) rotate(45deg)',
    },
    bottomRight: {
      top: arrowDistance,
      right: `${arrowOffset.x}`,
      left: 'auto',
      bottom: 'auto',
      transform: 'translate(0, -100%) rotate(45deg)',
    },
    left: {
      top: '50%',
      right: arrowDistance,
      left: 'auto',
      bottom: 'auto',
      transform: 'translate(100%, -50%) rotate(45deg)',
    },
    leftTop: {
      top: `${arrowOffset.y}`,
      right: arrowDistance,
      left: 'auto',
      bottom: 'auto',
      transform: 'translate(100%, -50%) rotate(45deg)',
    },
    leftBottom: {
      top: 'auto',
      right: arrowDistance,
      left: 'auto',
      bottom: `${arrowOffset.y}`,
      transform: 'translate(100%, 50%) rotate(45deg)',
    },
    right: {
      top: '50%',
      right: 'auto',
      left: arrowDistance,
      bottom: 'auto',
      transform: 'translate(-100%, -50%) rotate(45deg)',
    },
    rightTop: {
      top: `${arrowOffset.y}`,
      right: 'auto',
      left: arrowDistance,
      bottom: 'auto',
      transform: 'translate(-100%, -50%) rotate(45deg)',
    },
    rightBottom: {
      top: 'auto',
      right: 'auto',
      left: arrowDistance,
      bottom: `${arrowOffset.y}`,
      transform: 'translate(-100%, 50%) rotate(45deg)',
    },
  };

  return placements[placement] || placements['top'];
};
