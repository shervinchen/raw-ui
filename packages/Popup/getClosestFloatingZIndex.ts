import { isNaN } from 'lodash';

const FLOATING_CONTAINERS = ['raw-modal', 'raw-drawer', 'raw-popover'];

export const getClosestFloatingContainer = (
  element: HTMLElement | null
): HTMLElement | null => {
  if (!element || element === document.body) return null;

  const isFloatingContainer = FLOATING_CONTAINERS.some((container) =>
    element.classList.contains(container)
  );

  if (isFloatingContainer) {
    const isFloatingContainerWithPopup =
      element.parentElement.classList.contains('raw-popup');
    const floatingContainer = isFloatingContainerWithPopup
      ? element.parentElement
      : element;

    return floatingContainer;
  }

  return getClosestFloatingContainer(element.parentElement);
};

export const getZIndexByClosestFloating = (
  element: HTMLElement | null,
  defaultZIndex: number
) => {
  const floatingContainer = getClosestFloatingContainer(element);
  const zIndex = parseInt(
    floatingContainer !== null
      ? window.getComputedStyle(floatingContainer).zIndex
      : null,
    10
  );
  const closestZIndex = isNaN(zIndex) ? null : zIndex;
  if (closestZIndex !== null) {
    return closestZIndex + 1;
  }
  return defaultZIndex;
};
