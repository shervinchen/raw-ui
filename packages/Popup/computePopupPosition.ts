/* istanbul ignore file */
import { MutableRefObject } from 'react';
import { PopupPosition } from './Popup.types';

const getElementOffset = (element?: HTMLElement | null) => {
  if (!element) return { offsetTop: 0, offsetLeft: 0 };
  const { top, left } = element.getBoundingClientRect();
  return { offsetTop: top, offsetLeft: left };
};

export const computePopupPosition = (
  targetRef?: MutableRefObject<HTMLElement | null>,
  getContainer?: () => HTMLElement | null
): PopupPosition => {
  const targetRect = targetRef?.current?.getBoundingClientRect() ?? null;
  const bodyRect = document.body.getBoundingClientRect();
  const container = getContainer?.() ?? null;
  const { offsetTop, offsetLeft } = getElementOffset(container);

  if (!targetRect) {
    return {
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    };
  }

  return {
    top: container
      ? targetRect.top + container.scrollTop - offsetTop
      : targetRect.top - bodyRect.top,
    bottom: container
      ? targetRect.bottom + container.scrollTop - offsetTop
      : targetRect.bottom - bodyRect.top,
    left: container
      ? targetRect.left + container.scrollLeft - offsetLeft
      : targetRect.left - bodyRect.left,
    right: container
      ? targetRect.right + container.scrollLeft - offsetLeft
      : targetRect.right - bodyRect.left,
  };
};
