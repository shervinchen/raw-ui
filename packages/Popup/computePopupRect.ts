/* istanbul ignore file */
import { MutableRefObject } from 'react';
import { PopupRect } from './Popup.types';

const getElementOffset = (element?: HTMLElement | null) => {
  if (!element) return { offsetTop: 0, offsetLeft: 0 };
  const { top, left } = element.getBoundingClientRect();
  return { offsetTop: top, offsetLeft: left };
};

export const computePopupRect = (
  targetRef?: MutableRefObject<HTMLElement | null>,
  getContainer?: () => HTMLElement | null
): PopupRect => {
  const targetRect = targetRef?.current?.getBoundingClientRect() ?? null;
  const bodyRect = document.body.getBoundingClientRect();
  const container = getContainer?.() ?? null;
  const { offsetTop, offsetLeft } = getElementOffset(container);

  if (!targetRect) {
    return {
      top: 0,
      left: 0,
      width: 0,
    };
  }

  return {
    top: container
      ? targetRect.bottom + container.scrollTop - offsetTop
      : targetRect.bottom - bodyRect.top,
    left: container
      ? targetRect.left + container.scrollLeft - offsetLeft
      : targetRect.left - bodyRect.left,
    width: targetRect.width || targetRect.right - targetRect.left,
  };
};
