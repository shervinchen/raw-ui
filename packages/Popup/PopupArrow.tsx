import { FC, useMemo } from 'react';
import classNames from 'classnames';
import { getTargetRect, computePopupArrowPosition } from './computePopup';
import { PopupArrowProps, PopupArrowOffset } from './Popup.types';

const PopupArrow: FC<PopupArrowProps> = ({
  targetRef,
  placement,
  withBorder,
  className = '',
}) => {
  const arrowOffset = useMemo<PopupArrowOffset>(() => {
    const { width, height } = getTargetRect(targetRef);
    return {
      x: `${width / 2}px`,
      y: `${height / 2}px`,
    };
  }, [targetRef]);
  const { left, top, right, bottom, transform } = computePopupArrowPosition(
    arrowOffset,
    withBorder ? '4px' : '3px',
    placement
  );
  const classes = classNames('raw-popup-arrow', className);

  return (
    <span className={classes}>
      <style jsx>{`
        .raw-popup-arrow {
          width: 6px;
          height: 6px;
          position: absolute;
          left: ${left};
          top: ${top};
          right: ${right};
          bottom: ${bottom};
          transform: ${transform};
        }
      `}</style>
    </span>
  );
};

export default PopupArrow;
