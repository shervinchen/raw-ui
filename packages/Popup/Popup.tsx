import React, {
  FC,
  PropsWithChildren,
  useEffect,
  useState,
  MouseEvent,
  useCallback,
} from 'react';
import { createPortal } from 'react-dom';
import { PopupProps, PopupRect } from './Popup.types';
import {
  useClickAnyWhere,
  useMutationObserver,
  usePortal,
  useResize,
} from '../utils/hooks';
import { computePopupRect } from './computePopupRect';

const Popup: FC<PropsWithChildren<PopupProps>> = ({
  name,
  visible,
  targetRef,
  getPopupContainer,
  children,
}) => {
  const portal = usePortal(name, getPopupContainer);
  const [popupRect, setPopupRect] = useState<PopupRect>({
    top: 0,
    left: 0,
    width: 0,
  });

  const clickHandler = (event: MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    event.nativeEvent.stopImmediatePropagation();
    event.preventDefault();
  };

  const mouseDownHandler = (event: MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
  };

  const updatePopupRect = useCallback(() => {
    const rect = computePopupRect(targetRef, getPopupContainer);
    setPopupRect(rect);
  }, [getPopupContainer, targetRef]);

  useResize(updatePopupRect);

  useClickAnyWhere(() => {
    const { top, left } = computePopupRect(targetRef, getPopupContainer);
    /* istanbul ignore next */
    const shouldUpdate = top !== popupRect.top || left !== popupRect.left;
    /* istanbul ignore next */
    if (shouldUpdate) updatePopupRect();
  });

  useMutationObserver(targetRef, updatePopupRect);

  useEffect(() => {
    const targetNode = targetRef?.current ?? null;
    if (!targetNode) return;
    targetNode.addEventListener('mouseenter', updatePopupRect);
    /* istanbul ignore next */
    return () => {
      if (!targetNode) return;
      targetNode.removeEventListener('mouseenter', updatePopupRect);
    };
  }, [targetRef, updatePopupRect]);

  if (!targetRef) return null;
  if (!portal) return null;

  return createPortal(
    visible ? (
      <div
        className="raw-popup"
        onClick={clickHandler}
        onMouseDown={mouseDownHandler}
        data-testid="popup"
      >
        {children}
        <style jsx>{`
          .raw-popup {
            position: absolute;
            top: ${popupRect.top + 2}px;
            left: ${popupRect.left}px;
            width: ${popupRect.width}px;
          }
        `}</style>
      </div>
    ) : null,
    portal
  );
};

export default Popup;
