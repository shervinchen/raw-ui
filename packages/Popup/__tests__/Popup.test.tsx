import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Popup from '..';
import { MutableRefObject, useRef } from 'react';

const mockGetPopupPosition = jest.fn().mockReturnValue({
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
  transform: 'translate(0, 0)',
});

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

const Component = ({
  visible,
  targetRef,
  getPopupContainer,
}: {
  visible: boolean;
  targetRef?: MutableRefObject<HTMLElement | null>;
  getPopupContainer: () => HTMLElement | null;
}) => {
  return (
    <Popup
      name="testPopup"
      visible={visible}
      targetRef={targetRef}
      getPopupPosition={mockGetPopupPosition}
      getPopupContainer={getPopupContainer}
    >
      Popup Content
    </Popup>
  );
};

describe('Popup', () => {
  const targetRefMock = {
    current: document.createElement('div'),
  };
  const getPopupContainerMock = jest.fn();

  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render correctly content when visible is true', () => {
    render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(screen.queryByTestId('popup')).toBeInTheDocument();
  });

  test('should not render content when visible is false', () => {
    render(
      <Component
        visible={false}
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(screen.queryByTestId('popup')).not.toBeInTheDocument();
  });

  test('should update popup position when targetRef changes', () => {
    const { rerender } = render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    const newTargetRefMock = {
      current: document.createElement('div'),
    };
    rerender(
      <Component
        visible
        targetRef={newTargetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
  });

  test('should update popup position when window resize', () => {
    render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    fireEvent(window, new Event('resize'));
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(2);
  });

  test('should update popup position when mouse over', async () => {
    render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    await user.hover(targetRefMock.current);
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(2);
  });

  test('should update popup position when document size change', async () => {
    render(
      <Component visible targetRef={targetRefMock} getPopupContainer={null} />
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    document.body.style.height = '10000px';
    await waitFor(() => {
      expect(mockGetPopupPosition).toHaveBeenCalledTimes(2);
    });
  });

  test('should update popup position when container size change', async () => {
    render(
      <div
        id="parentElement"
        style={{
          position: 'relative',
          overflowY: 'auto',
          width: '400px',
          height: '200px',
        }}
      >
        <Component
          visible
          targetRef={targetRefMock}
          getPopupContainer={() => document.querySelector('#parentElement')}
        />
      </div>
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    const parentElement = document.querySelector(
      '#parentElement'
    ) as HTMLElement;
    parentElement.style.height = '10000px';
    await waitFor(() => {
      expect(mockGetPopupPosition).toHaveBeenCalledTimes(2);
    });
  });

  test('should render to specified container', () => {
    const Container = () => {
      const ref = useRef<HTMLDivElement>(null);
      const customContainerRef = useRef<HTMLDivElement>(null);

      return (
        <div>
          <div ref={customContainerRef} data-testid="customContainer" />
          <div ref={ref}>
            <Component
              visible
              targetRef={ref}
              getPopupContainer={() => customContainerRef.current}
            />
          </div>
        </div>
      );
    };
    render(<Container />);
    expect(
      screen
        .getByTestId('customContainer')
        .contains(screen.getByTestId('popup'))
    ).toBeTruthy();
  });

  test('should prevent event on popup', async () => {
    render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    const popup = screen.getByTestId('popup');
    await user.click(popup);
    expect(popup).toBeInTheDocument();
  });

  test('should not trigger mouse over event when target is empty', async () => {
    render(<Component visible getPopupContainer={getPopupContainerMock} />);
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    await user.hover(targetRefMock.current);
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
  });

  test('should not trigger mouse over event when event is removed', async () => {
    const { unmount } = render(
      <Component
        visible
        targetRef={targetRefMock}
        getPopupContainer={getPopupContainerMock}
      />
    );
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
    unmount();
    await user.hover(targetRefMock.current);
    expect(mockGetPopupPosition).toHaveBeenCalledTimes(1);
  });
});
