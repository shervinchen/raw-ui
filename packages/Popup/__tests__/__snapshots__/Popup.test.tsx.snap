// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should match the snapshot 1`] = `<DocumentFragment />`;
